src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs

```diff
@@ -400,10 +404,24 @@ public string GetAbsolutePath(string executingFilePath, string pagePath)
             ViewLocationExpanderContext expanderContext,
             ViewLocationCacheKey cacheKey)
         {
-            // Only use the area view location formats if we have an area token.
-            IEnumerable<string> viewLocations = !string.IsNullOrEmpty(expanderContext.AreaName) ?
-                _options.AreaViewLocationFormats :
-                _options.ViewLocationFormats;
+            IEnumerable<string> viewLocations;
+            if (!string.IsNullOrEmpty(expanderContext.AreaName) && 
+                !string.IsNullOrEmpty(expanderContext.ControllerName))
+            {
+                viewLocations = _options.AreaViewLocationFormats;
+            }
+            else if (!string.IsNullOrEmpty(expanderContext.ControllerName))
+            {
+                viewLocations = _options.ViewLocationFormats;
+            }
+            else if (!string.IsNullOrEmpty(expanderContext.PageName))
+            {
+                viewLocations = _options.PageViewLocationFormats;
+            }
+            else
+            {
+                throw null;
+            }
```

Comment:
I'm not super thrilled about this coupling, but I didn't want to implement a new abstraction.